src/instruction_graph_generator.cc

```diff
@@ -1749,57 +1749,57 @@ void generator_impl::compile_execution_command(batch& command_batch, const execu
-	// command that generated the pushed data. This will allow computation-communication overlap, especially in the case of oversubscribed splits.
-	std::vector<region<3>> concurrent_send_regions;
-	// Since we now send boxes individually, we do not need to allocate the entire push_box contiguously.
-	box_vector<3> required_host_allocation;
-	{
-		std::unordered_map<instruction_id, box_vector<3>> individual_send_boxes;
-		for(auto& [box, original_writer] : buffer.original_writers.get_region_values(push_box)) {
-			individual_send_boxes[original_writer->get_id()].push_back(box);
-			required_host_allocation.push_back(box);
+	for(const auto& [target, reg] : pcmd.get_target_regions()) {
+		// If not all nodes contribute partial results to a global reductions, the remaining ones need to notify their peers that they should not expect any
+		// data. This is done by announcing an empty box through the pilot message, but not actually performing a send.
+		if(reg.empty()) {
+			assert(trid.rid != no_reduction_id);
+			create_outbound_pilot(command_batch, target, trid, box<3>());
+			continue;
 		}
-		for(auto& [original_writer, boxes] : individual_send_boxes) {
-			concurrent_send_regions.push_back(region(std::move(boxes)));
+
+		// Prioritize all instructions participating in a "push" to hide the latency of establishing local coherence behind the typically much longer latencies
+		// of inter-node communication
+		command_batch.base_priority = 10;
+
+		auto& buffer = m_buffers.at(trid.bid);
+		auto& host_memory = buffer.memories[host_memory_id];
+
+		// We want to generate the fewest number of send instructions possible without introducing new synchronization points between chunks of the same
+		// command that generated the pushed data. This will allow computation-communication overlap, especially in the case of oversubscribed splits.
+		std::vector<region<3>> concurrent_send_regions;
+		// Since we now send boxes individually, we do not need to allocate the entire push_box contiguously.
+		box_vector<3> required_host_allocation;
+		{
+			std::unordered_map<instruction_id, box_vector<3>> individual_send_boxes;
+			for(auto& [box, original_writer] : buffer.original_writers.get_region_values(reg)) {
+				individual_send_boxes[original_writer->get_id()].push_back(box);
+				required_host_allocation.push_back(box);
+			}
+			for(auto& [original_writer, boxes] : individual_send_boxes) {
+				concurrent_send_regions.push_back(region(std::move(boxes)));
+			}
 		}
-	}
 
-	allocate_contiguously(command_batch, trid.bid, host_memory_id, std::move(required_host_allocation));
-	establish_coherence_between_buffer_memories(command_batch, trid.bid, host_memory_id, concurrent_send_regions);
+		allocate_contiguously(command_batch, trid.bid, host_memory_id, std::move(required_host_allocation));
+		establish_coherence_between_buffer_memories(command_batch, trid.bid, host_memory_id, concurrent_send_regions);
```

Comment:
These must be outside the loop, otherwise we get reallocations / staging copies for every target node instead of once

Reply:
Out of scope for this PR, as discussed below.